crypto/ui/ui_lib.c

```diff
@@ -387,10 +390,38 @@ char *UI_construct_prompt(UI *ui, const char *object_desc,
 void *UI_add_user_data(UI *ui, void *user_data)
 {
     void *old_data = ui->user_data;
+
+    if ((ui->flags & UI_FLAG_DUPL_DATA) != 0) {
+        ui->meth->ui_destroy_data(ui, old_data);
+        old_data = NULL;
+    }
     ui->user_data = user_data;
+    ui->flags &= ~UI_FLAG_DUPL_DATA;
     return old_data;
 }
 
+int UI_dup_user_data(UI *ui, void *user_data)
```

Comment:
That would be inconsistent with the rest of the UI API.  You might want to argue that the API should be "modernised", but that's for 1.2.0, not for now.

Background: the whole set0 / set1 semantics came after UI, as far as I recall, or was at least not settled and well known throughout the whole team at the time.